packages/gateway/src/Shard.ts

```diff
@@ -339,17 +427,74 @@ export class DiscordenoShard {
 
   /** Handle an incoming gateway message. */
   async handleMessage(message: MessageEvent): Promise<void> {
-    let preProcessMessage = message.data
+    // The ws npm package will use a Buffer, while the global built-in will use ArrayBuffer
+    const isCompressed = message.data instanceof ArrayBuffer || message.data instanceof Buffer
+
+    const data = isCompressed ? await this.decompressPacket(message.data) : (JSON.parse(message.data) as DiscordGatewayPayload)
+
+    // Check if the decompression was not successful
+    if (!data) return
 
-    // If message compression is enabled, Discord might send zlib compressed payloads.
-    if (this.gatewayConfig.compress && preProcessMessage instanceof Blob) {
-      preProcessMessage = inflateSync(await preProcessMessage.arrayBuffer()).toString()
+    await this.handleDiscordPacket(data)
+  }
+
+  /**
+   * Decompress a zlib/zstd compressed packet
+   *
+   * @private
+   */
+  async decompressPacket(data: ArrayBuffer | Buffer): Promise<DiscordGatewayPayload | null> {
+    // A buffer is a Uint8Array under the hood. An ArrayBuffer is generic, so we need to create the Uint8Array that uses the whole ArrayBuffer
+    const compressedData: Uint8Array = data instanceof Buffer ? data : new Uint8Array(data)
+
+    if (this.gatewayConfig.transportCompression === TransportCompression.zlib) {
+      if (!this.inflate) {
+        this.logger.fatal('[Shard] zlib-stream transport compression was enabled. But no instance of Inflate was found.')
+        return null
+      }
+
+      // Alias, used to avoid some null checks in the Promise constructor
```

Comment:
Why not `this.inflate?.` or `this.inflate!.`?

Reply:
~~`?.` is runtime and `!.` is lying to TS *tecnically*~~

If you think it is clearer fine, ig they are used more often rather than aliasing to avoid these issues